src/Interpreters/ActionsDAG.cpp

```diff
@@ -1941,4 +1942,102 @@ ActionsDAGPtr ActionsDAG::cloneActionsForFilterPushDown(
     return actions;
 }
 
+static bool chainPreservesSorting(const ActionsDAG::Node* chain)
+{
+    const Field field{};
+    const ActionsDAG::Node* node = chain;
+    while (node)
+    {
+        if (node->type == ActionsDAG::ActionType::FUNCTION)
+        {
+            auto func = node->function_base;
+            if (func)
+            {
+                if (!func->hasInformationAboutMonotonicity())
+                    return false;
+
+                const auto & types = func->getArgumentTypes();
+                if (types.empty())
+                    return false;
+
+                /// TODO: we support monotonicity check only for functions with one parameter but ...
+                ///       if one parameter is variable and other are constant then we can try to check monotonicity as well
+                const auto monotonicity = func->getMonotonicityForRange(*types.front(), field, field);
+                if (!monotonicity.is_always_monotonic)
+                    return false;
+            }
+        }
+
+        if (node->children.empty())
+            break;
+
+        node = node->children.front();
+    }
+    return true;
+}
+
+static const ActionsDAG::Node* findColumn(const ActionsDAG::Node * start_node, const String & sorted_column)
+{
+    const ActionsDAG::Node * current = start_node;
+    while (current)
+    {
+        /// if column found
+        if (current->type == ActionsDAG::ActionType::INPUT && current->result_name == sorted_column)
+            return current;
+
+        if (current->children.empty())
+            break;  /// column not found
+
+        current = current->children.front();
+    }
+    return nullptr;
+}
+
+bool ActionsDAG::isSortingPreserved(
+    const Block & input_header, const SortDescription & sort_description, const String & ignore_output_column) const
+{
+    if (sort_description.empty())
+        return true;
+
+    if (hasArrayJoin())
+        return false;
+
+    const Block & output_header = updateHeader(input_header);
+    for (const auto & desc : sort_description)
+    {
+        /// header contains column with the same name
+        if (output_header.findByName(desc.column_name))
+        {
+            /// find the corresponding node in output
+            const auto * output_node = tryFindInOutputs(desc.column_name);
+            if (!output_node)
+            {
+                /// sorted column name in header but NOT in expression output -> no expression is applied to it -> sorting preserved
+                continue;
+            }
+        }
```

Comment:
This code is ok.
I think it is the same as to check if there is an input with this name.